Shell completion of dependencies and executables

Bouke opened the pull request:
This PR enables shell completions for dependencies:

```
$ swift package edit A[tab]
$ swift package edit Alamofire
```

and executables:

```
$ swift run [tab]
$ swift run lark-generate-client
```

aciidb0mb3r commented on Sources/Commands/Completions+zsh.swift:
```diff
@@ -12,6 +12,9 @@ import Foundation
 import Utility
 
+let dependencies = "swift package show-dependencies | grep -oE '── (.*)<' | sed -E 's/── (.*)</\\1/'"
+let executables = "swift package describe | grep 'Type: executable' -B 3 | grep 'Name:' | sed -e 's/    Name: //'"
```
Instead of shelling out to get this data, we can directly get it from the package graph object.

Bouke replied:
Can you show me what that would look like?

aciidb0mb3r replied:
Ah, I just realized this is dynamic and needs to run per package. Maybe the best thing to do is add a dedicated subcommand for autocompletion, so we don't need to run grep and sed. 
Something like: 
`swift package shell-autocomplete --list-executables`
`swift package shell-autocomplete --list-dependencies`

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -200,6 +203,26 @@ public struct PathArgument: ArgumentKind {
     public static var completion: ShellCompletion = .filename
 }
 
+public struct ExecutableName: ArgumentKind {
```
I think these do not belong here because they're not generic enough for the base argument parser library.

Bouke replied:
Yes, they are better suited for the Commands module.

aciidb0mb3r commented on Sources/Utility/ArgumentParser.swift:
```diff
@@ -200,6 +203,26 @@ public struct PathArgument: ArgumentKind {
+        name = argument
+    }
+
+    public static var completion: ShellCompletion = .function("_swift_executable")
```
The two new structs are doing same thing as a StringArgument, so we should probably find a way to use that. Maybe we can make `ShellCompletion` an instance variable instead of static?

Bouke replied:
It needs to work with instances of `ArgumentProtocol`, so an instance variable will not work. We could add the completion to instances of `ArgumentProtocol`, but than the coupling between the ArgumentKind and it’s completions isn’t explicit.

I designed the current implementation such that the ArgumentKind is responsible for providing it’s completions. That’s also why I think that adding two `ArgumentKind`s is no such a big deal, but rather a feature — making the argument’s kind even more explicit.

aciidb0mb3r replied:
I think it actually make sense to move it to `ArgumentProtocol` and remove the coupling from the type. That way we can add completion option when constructing an option. This will allow us to have greater flexibility as each option can then decide if they want autocompletion and of what kind.

Bouke commented:
I've updated the PR. I've added the ability for arguments to override the type's completion. I've left the current completions based on `ArgumentKind.completion` intact -- this allows for a fallback; e.g. `PathArgument`. I'll be unable to respond for the next week or two, feel free to merge it in with alterations.

Bouke commented:
Bump

aciidb0mb3r commented on Sources/Commands/SwiftPackageTool.swift:
```diff
@@ -174,7 +174,12 @@ public class SwiftPackageTool: SwiftTool<PackageToolOptions> {
         case .describe:
             let graph = try loadPackageGraph()
-            describe(graph.rootPackages[0].underlyingPackage, in: options.describeMode, on: stdoutStream)
+            switch options.describeExecutableNamesOnly {
```
Instead of piggybacking on existing modes, we should add a new mode specifically for shell autocompletion. It will be easier to maintain and extend in future. Example from my previous comment:

```
swift package shell-autocomplete --list-executables
swift package shell-autocomplete --list-dependencies
```

Bouke replied:
I've refactored the code to this. Please re-review.

Bouke commented on Sources/Commands/SwiftPackageTool.swift:
```diff
@@ -310,25 +321,32 @@ public class SwiftPackageTool: SwiftTool<PackageToolOptions> {
+            option: completionToolParser.add(
+                option: "--list-executables", kind: Bool.self,
+                usage: "List all executables"),
+            to: { $0.completionToolMode = .listExecutables; _ = $1 })
```
Is there a cleaner way to say we're only interested in `$0` without explicitly naming them?

aciidb0mb3r replied:
Ah, we should make the kind here to be `CompletionToolMode`. See `initMode` for example implementation.

aciidb0mb3r replied:
Not sure but we may need to flatten the `Shell` enum for this to work.

```
enum CompletionToolMode {
    case generateShellScriptBash
    case generateShellScriptZSH
    case listDependencies
    case listExecutables
}
```

Bouke commented on Sources/Commands/SwiftPackageTool.swift:
```diff
@@ -412,7 +437,7 @@ public enum PackageMode: String, StringEnumArgument {
     case fetch
     case generateXcodeproj = "generate-xcodeproj"
-    case generateCompletionScript = "generate-completion-script"
+    case completionTool = "completion-tool"
```
I've named it `completion-tool` instead of `shell-autocomplete` as it is technically not autocompleting. Only suggestion completions when asked.

aciidb0mb3r commented on Sources/Commands/Describe.swift:
```diff
@@ -29,6 +29,14 @@ func describe(_ package: Package, in mode: DescribeMode, on stream: OutputByteSt
     stream.flush()
 }
 
+func describeExecutableNames(_ package: Package, on stream: OutputByteStream) {
```
I think we can just inline the contents of this method.

Bouke replied:
Ok, changed.

Bouke commented:
I've made the requested changes, seems like it's good to go now?

aciidb0mb3r commented:
Yep, this looks good. Thanks a lot for working on this @Bouke!

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
@shahmishal For some reason I am unable to kick off swift-ci on this repo

aciidb0mb3r commented:
@swift-ci please smoke test

Bouke commented:
@aciidb0mb3r @shahmishal Any ideas on getting swift-ci running for this PR?

shahmishal commented:
I will look into it today

Bouke commented:
Any ideas yet?

aciidb0mb3r commented:
@swift-ci please smoke test

aciidb0mb3r commented:
I've filed a bug: https://bugs.swift.org/browse/SR-5681

Bouke commented:
Can you merge the PR after running the test on your local machine?

aciidb0mb3r commented:
No, that is not allowed because it might cause issues in the ecosystem and block PRs on other projects.

Bouke commented:
The bug was marked as resolved, can you try again?

aciidb0mb3r commented:
@swift-ci please smoke test
